api/build/graph.go

```diff
@@ -0,0 +1,577 @@
+	"github.com/go-vela/server/router/middleware/org"
+	"github.com/go-vela/server/router/middleware/repo"
+	"github.com/go-vela/server/router/middleware/user"
+	"github.com/go-vela/server/scm"
+	"github.com/go-vela/server/util"
+	"github.com/go-vela/types"
+	"github.com/go-vela/types/constants"
+	"github.com/go-vela/types/library"
+	"github.com/go-vela/types/pipeline"
+	"github.com/sirupsen/logrus"
+)
+
+// Graph contains nodes, and relationships between nodes, or edges.
+//
+//	a node is a pipeline stage and its relevant steps.
+//	an edge is a relationship between nodes, defined by the 'needs' tag.
+//
+// swagger:model Graph
+type Graph struct {
+	BuildID int64         `json:"build_id"`
+	Nodes   map[int]*node `json:"nodes"`
+	Edges   []*edge       `json:"edges"`
+}
+
+// node represents a pipeline stage and its relevant steps.
+type node struct {
+	ID      int    `json:"id"`
+	Cluster int    `json:"cluster"`
+	Name    string `json:"name"`
+
+	// vela metadata
+	Status     string          `json:"status"`
+	StartedAt  int             `json:"started_at"`
+	FinishedAt int             `json:"finished_at"`
+	Steps      []*library.Step `json:"steps"`
+
+	// unexported data used for building edges
+	Stage *pipeline.Stage `json:"-"`
+}
+
+// edge represents a relationship between nodes, primarily defined by a stage 'needs' tag.
+type edge struct {
+	Cluster     int `json:"cluster"`
+	Source      int `json:"source"`
+	Destination int `json:"destination"`
+
+	// vela metadata
+	Status string `json:"status"`
+}
+
+// stg represents a stage's steps and some metadata for producing node/edge information
+type stg struct {
+	steps []*library.Step
+	// used for tracking stage status
+	success    int
+	running    int
+	failure    int
+	killed     int
+	canceled   int
+	skipped    int
+	startedAt  int
+	finishedAt int
+}
+
+const (
+	// clusters determine graph orientation
+	BuiltInCluster       = 2
+	PipelineCluster      = 1
+	ServiceCluster       = 0
+	GraphComplexityLimit = 1000 // arbitrary value to limit render complexity
+)
+
+// swagger:operation GET /api/v1/repos/{org}/{repo}/builds/{build}/graph builds GetBuildGraph
+//
+// Get directed a-cyclical graph for a build in the configured backend
+//
+// ---
+// produces:
+// - application/json
+// parameters:
+// - in: path
+//   name: org
+//   description: Name of the org
+//   required: true
+//   type: string
+// - in: path
+//   name: repo
+//   description: Name of the repo
+//   required: true
+//   type: string
+// - in: path
+//   name: build
+//   description: Build number
+//   required: true
+//   type: integer
+// security:
+//   - ApiKeyAuth: []
+// responses:
+//   '200':
+//     description: Successfully retrieved graph for the build
+//     schema:
+//       type: array
+//       items:
+//         "$ref": "#/definitions/Graph"
+//   '500':
+//     description: Unable to retrieve graph for the build
+//     schema:
+//       "$ref": "#/definitions/Error"
```

Comment:
I believe this endpoint also can return 401 and 404?

Reply:
true! adjusting